### PR TITLE
fix(CompilerCli): assert that all pipes and directives are declared

### PR DESCRIPTION
**Pull request by vicb**

### CODE REVIEW

File: modules/@angular/compiler/src/offline_compiler.ts [hunk 1]
```diff
@@ -29,55 +29,93 @@ export class SourceModule {
...
-  ngModuleByDirective: Map<StaticSymbol, CompileNgModuleMetadata>,
+    programStaticSymbols: StaticSymbol[], options: {transitiveModules: boolean},
+    metadataResolver: CompileMetadataResolver): {
+  ngModuleByPipeOrDirective: Map<StaticSymbol, CompileNgModuleMetadata>,
   files: Array<{srcUrl: string, directives: StaticSymbol[], ngModules: StaticSymbol[]}>
 } {
+  const programNgModules: CompileNgModuleMetadata[] = [];
+  const programPipesOrDirectives: StaticSymbol[] = [];
+
+  // first split all the symbols from the programs in 2 buckets
```
Comment (tbosch) [inline modules/@angular/compiler/src/offline_compiler.ts]: Maybe create helper function for this?

File: modules/@angular/compiler/src/offline_compiler.ts [hunk 2]
```diff
@@ -134,7 +174,7 @@ export class OfflineCompiler {
           if (!compMeta.isComponent) {
             return Promise.resolve(null);
           }
-          const ngModule = ngModuleByDirective.get(dirType);
+          const ngModule = ngModuleByPipeOrDirective.get(dirType);
           if (!ngModule) {
             throw new Error(`Cannot determine the module for component ${compMeta.type.name}!`);
```
Comment (tbosch) [inline modules/@angular/compiler/src/offline_compiler.ts]: Change message to `Internal error: ...`

File: modules/@angular/compiler-cli/src/extractor.ts [hunk 3]
```diff
@@ -55,27 +55,22 @@ export class Extractor {
       }
 
       const staticType = this.reflectorHost.findDeclaration(absSourcePath, symbol, absSourcePath);
```
Comment (tbosch) [inline modules/@angular/compiler-cli/src/extractor.ts]: Try to reuse CodeGeneratorSymbolCollector

File: modules/@angular/compiler/src/offline_compiler.ts [hunk 4]
```diff
@@ -29,56 +29,81 @@ export class SourceModule {
...
 
+  // Throw an error if any of the program pipe or directives is not declared by a module
+  const symbolsMissingModule =
+      programPipesOrDirectives.filter(s => !ngModuleByPipeOrDirective.has(s));
+
+  if (symbolsMissingModule.length) {
+    const messages = symbolsMissingModule.map(
+        s => `Cannot determine the module for class ${s.name} in ${s.filePath}!`);
+    throw new Error(messages.join('\n'));
+  }
```
Comment (vicb) [inline modules/@angular/compiler/src/offline_compiler.ts]: accumulating errors to ease fixing (before we were throwing on the first error)

**Comment by vicb** [comment]
@tbosch PTAL, I have added inline comments on things I have updated since your last review. Thanks.

File: modules/@angular/compiler/src/offline_compiler.ts [hunk 5]
```diff
@@ -29,56 +29,81 @@ export class SourceModule {
...
+      ngModuleByPipeOrDirective.set(pipeMeta.type.reference, ngModuleMeta);
     });
   });
 
+  // Throw an error if any of the program pipe or directives is not declared by a module
+  const symbolsMissingModule =
+      programPipesOrDirectives.filter(s => !ngModuleByPipeOrDirective.has(s));
+
+  if (symbolsMissingModule.length) {
+    const messages = symbolsMissingModule.map(
```
Comment (tbosch) [inline modules/@angular/compiler/src/offline_compiler.ts]: Do a local run to see if the error message looks ok.

Comment (vicb) [inline modules/@angular/compiler/src/offline_compiler.ts]:
```
Cannot determine the module for class FailCmp in /private/var/folders/ns/9rf1lry94vd1w2x7c_9lfp08009nrm/T/e2e_test.1477497785/src/entry_components.ts!
Cannot determine the module for class FailDirective in /private/var/folders/ns/9rf1lry94vd1w2x7c_9lfp08009nrm/T/e2e_test.1477497785/src/entry_components.ts!
```

**Pull request merged**